include/tvm/runtime/container.h

```diff
@@ -189,6 +189,759 @@ class InplaceArrayBase {
+    if (data_ == nullptr) {
+      SwitchContainer(n);
+      return;
+    }
+    int64_t size = GetArrayNode()->size_;
+    if (size < n) {
+      CopyOnWrite(n - size)->Enlarge(n - size);
+    } else if (size > n) {
+      CopyOnWrite()->Shrink(size - n);
+    }
+  }
+
+  /*!
+   * \brief Make sure the list has the capacity of at least n
+   * \param n lower bound of the capacity
+   */
+  void reserve(int64_t n) {
+    if (data_ == nullptr || n > GetArrayNode()->capacity_) {
+      SwitchContainer(n);
+    }
+  }
+
+  /*! \brief Release reference to all the elements */
+  void clear() {
+    if (data_ != nullptr) {
+      ArrayNode* p = CopyOnWrite();
+      p->clear();
+    }
+  }
+
+ public:
+  // Array's own methods
+
+  /*!
+   * \brief set i-th element of the array.
+   * \param i The index
+   * \param value The value to be setted.
+   */
+  void Set(int64_t i, const T& value) {
+    ArrayNode* p = this->CopyOnWrite();
+    (*p)[i] = value;
+  }
+
+  /*! \return The underlying ArrayNode */
+  ArrayNode* GetArrayNode() const { return static_cast<ArrayNode*>(data_.get()); }
+
+  /*!
+   * \brief Helper function to apply fmutate to mutate an array.
+   * \param fmutate The transformation function T -> T.
+   * \tparam F the type of the mutation function.
+   * \note This function performs copy on write optimization.
+   */
+  template <typename F>
+  void MutateByApply(F fmutate) {
+    if (data_ == nullptr) {
+      return;
+    }
+    ArrayNode* p = GetArrayNode();
+    ObjectRef* iter = p->MutableBegin();
+    if (data_.unique()) {
+      for (int64_t i = 0; i < p->size_; ++i) {
+        T old_elem = DowncastNoCheck<T>(std::move(*iter));
+        T new_elem = fmutate(std::move(old_elem));
+        *iter++ = std::move(new_elem);
+      }
+    } else {
+      for (int64_t i = 0; i < p->size_; ++i) {
+        T old_elem = DowncastNoCheck<T>(*iter);
+        T new_elem = fmutate(old_elem);
+        // do nothing until the first real mutation begins.
+        if (new_elem.same_as(*iter++)) {
+          continue;
+        }
+        ObjectPtr<ArrayNode> copy = ArrayNode::Empty(p->capacity_);
+        // copy [0, i)
+        int64_t& size = copy->size_ = 0;
+        for (ObjectRef* old_iter = p->MutableBegin(); size < i;) {
+          copy->EmplaceInit(size++, *old_iter++);
+        }
+        // copy [i]
+        copy->EmplaceInit(size++, std::move(new_elem));
+        // complete transformation in (i, size)
+        while (size < p->size_) {
+          T old_elem = DowncastNoCheck<T>(*iter++);
+          T new_elem = fmutate(old_elem);
+          copy->EmplaceInit(size++, std::move(new_elem));
+        }
+        data_ = std::move(copy);
+        break;
+      }
+    }
+  }
+
+  /*!
+   * \brief reset the array to content from iterator.
+   * \param first begin of iterator
+   * \param last end of iterator
+   * \tparam IterType The type of iterator
+   */
+  template <typename IterType>
+  void Assign(IterType first, IterType last) {
```

Comment:
Can we do COW here instead of always create a new Array?

Reply:
e.g. reserve then AssignRange

Reply:
`reserve` doesn't necessarily provide a unique copy when `capacity >= required-capacity`. We can do CoW -> Shrink -> AssignRange instead

Reply:
Hmmm I think we should give a better name than `AssignRange`. `InitRange` sounds better

Reply:
I am eating my words. I think neither of the two solutions are problem-free.
1) reserve-and-AssignRange does not guarantee CoW
2) Cow-shrink-AssignRange fail to ensure exception safety in AssignRange because it doesn't change `size`

Reply:
COW shrink AssignRange seems to be OK for me. Given that size is changed in Shrink?

Reply:
BTW, the new logic LGTM

Reply:
I renamed `AssignRange` to `InitRange`, because the size_ is not changed so exception safety won't be preserved in this case.